app/Svg/Parser.hs

```diff
@@ -344,82 +367,106 @@ getTransform :: Tag T.Text -> Point
-      tokenize d' = map (T.splitOn ",") $ filter (not . T.null) $ T.splitOn " " d'
-
-      parseTokens :: [[T.Text]] -> [Maybe Point] -> [Maybe Point]
-      parseTokens [] points = points
-      -- TODO: Fix support for Bezier curves (indicated by the "C"/"c").
-      parseTokens (["C"]:_:_:d') points = parseTokens d' points
-      parseTokens (["c"]:_:_:d') points = parseTokens d' points
-
-      -- Close loop to beginning of path.
-      parseTokens [["Z"]] points = points ++ [head points]
-      parseTokens [["z"]] points = points ++ [head points]
-
-      -- Absolute move.
-      parseTokens (["L"]:d') points = parseTokens d' points
-      parseTokens (["M"]:d') points = parseTokens d' points
-
-      -- Relative move.
-      parseTokens (["m"]:[x,y]:d') points =
-        parseTokens (["m"]:d') $
-          points ++ [addTuples <$> convertToPoint (x,y) <*> (if null points then Just (0, 0) else last points)]
-      parseTokens (["m"]:d') points = parseTokens d' points
-      parseTokens ([x, y]:d') points = parseTokens d' $ points ++ [convertToPoint (x, y)]
-    --   parseTokens (["l"]:d') points = parseTokens d' points
-
-      -- Absolute horizontal/vertical move.
-      parseTokens (["H"]:[x2]:d') points =
-        parseTokens d' $ points ++ [(,) <$> readMaybe (T.unpack x2) <*> (snd <$> last points)]
-      parseTokens (["V"]:[y2]:d') points =
-        parseTokens d' $ points ++ [(,) <$> (fst <$> last points) <*> readMaybe (T.unpack y2)]
-
-      -- Relative horizontal/vertical move.
-      parseTokens (["h"]:[x2]:d') points =
-        parseTokens d' $ points ++ [(,) <$> ((+) <$> (fst <$> last points) <*> readMaybe (T.unpack x2)) <*> (snd <$> last points)]
-      parseTokens (["v"]:[y2]:d') points =
-        parseTokens d' $ points ++ [(,) <$> (fst <$> last points) <*> ((+) <$> (snd <$> last points) <*> readMaybe (T.unpack y2))]
-
-      -- Error case.
-      parseTokens d' _ = error (show d')
-
-      convertToPoint (x, y) =
-        case (readMaybe (T.unpack x), readMaybe (T.unpack y)) of
-            (Just m, Just n) -> Just (m, n)
-            _ -> Nothing
+        -- Keep track of the (first point, path mode, most recent point)
+        initialState = PathDState { firstPoint = (0, 0)
+                                  , mode = AbsoluteMove
+                                  , currentPoint = (0, 0) }
+
+        parser = do
+            p <- parseStep
+            -- Record the first point added.
+            _ <- P.modifyState $ \st -> st { firstPoint = p, currentPoint = p }
+            rest <- P.many parseStep
+            return (p : rest)
+        parseStep = do
+            _ <- (P.option () $ bezier <|>
+                absoluteMove <|> relativeMove <|>
+                absHorizontalMove <|> absVerticalMove <|>
+                relHorizontalMove <|> relVerticalMove) >> P.spaces
+            p <- closeLoop <|> moveArg
+            state <- P.getState
+            let (currentX, currentY) = currentPoint state
```

Comment:
I'd like you to move this part into a separate helper of type `Point -> PathDState -> PathDState ` (extract the relevant parts of the current state in the helper). Then you can do `let p' = (yourHelper p)` etc. in the main parser. Also see my above comment about `P.spaces`.